src/main/java/org/opensearch/knn/index/engine/ResolvedRequiredParameters.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.engine;
+
+import lombok.Getter;
+import org.opensearch.Version;
+import org.opensearch.common.Nullable;
+import org.opensearch.common.ValidationException;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.knn.index.SpaceType;
+import org.opensearch.knn.index.VectorDataType;
+import org.opensearch.knn.index.engine.config.CompressionConfig;
+import org.opensearch.knn.index.engine.config.WorkloadModeConfig;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import static org.opensearch.knn.index.mapper.KNNVectorFieldMapperUtil.createKNNMethodContextFromLegacy;
+
+/**
+ * Resolved parameters required for constructing a {@link KNNIndexContext}. If any of these parameters can be null,
+ * then their getters need to be wrapped in an {@link java.util.Optional}
+ */
+public final class ResolvedRequiredParameters {
+    @Getter
+    private final VectorDataType vectorDataType;
+    @Getter
+    private final WorkloadModeConfig mode;
+    @Getter
+    private final SpaceType spaceType;
+    @Getter
+    private final KNNEngine knnEngine;
+    @Getter
+    private final CompressionConfig compressionConfig;
+    @Getter
+    private final Version createdVersion;
+    @Getter
+    private final int dimension;
+    @Nullable
+    private final KNNMethodContext knnMethodContext;
+
+    /**
+     *
+     * @param originalParameters The original user provided parameters
+     * @param settings Settings for the index; passing null will mean that it is not possible to resolve for the legacy
+     * @param createdVersion version that this was created for
+     */
+    public ResolvedRequiredParameters(UserProvidedParameters originalParameters, Settings settings, Version createdVersion) {
```

Comment:
There are places in the code where we have to add logic in constructor, seems like we can avoid it here. In general logic in constructor is code smell

Consider a ParameterResolver class which has a function that takes these parameters and returns this object. This will make this pojo light weight. Not super urgent